Create basic specification for the translation section (#120)
PrestaShop/prestashop-specs: back-office/international/translations/translations.md

# Translations

## Conceptual vocabulary

- Wordings: original texts found throughout the software. They are in English but may be translated _to_ English for better localization.
- Translations: wordings translated to a given language.
- Translation domain: named logic group of wordings. Any given wording is attached to one and only one translation domain. Two identical wordings attached to two different translation domains are considered as two different wordings.
- Translation catalogue: A collection of translation domains containing translations for a given language, stored physically in a support like files or on a database.

> - Default catalogue: The master translation catalogue that contains the wordings but not its translations.

Author: ```suggestion
- Default catalogue: The master translation catalogue that contains the wordings but not its translations. It's [included in the GitHub project](https://github.com/PrestaShop/PrestaShop/tree/1.7.6.0/app/Resources/translations/default).
```

## Modify translations section

> This section can be used the user can modify all translations from PrestaShop.

Author: Oops, looks like I reworded this in my sleep 🤪 

```suggestion
This section allows the user to modify all translations used in PrestaShop.
```

In order to modify translations, the user must first choose which translation type they want to edit, then the language they want to edit it for (eg. French).

The translation types are as follows:

- Back office translations
- Front office translations
- Installed module translations
- Email translations
- Other translations

Once you click on the "Edit" button, you go to the translation interface.

### Particular cases

> - When the user chooses "Front office translations", a list of all the installed themes except "classic" + a special item called "Core translations" is displayed as well.

Author: Clarified below

#### Translation sources & storage

This section covers where wordings are sourced from and where translations are stored to according to the translation type chosen.

##### Generalities

Usually and unless specified otherwise:

- Wordings are sourced from XLF files stored somewhere, one file per translation domain.
- User-edited translations are stored in the database

##### Back office translations

Wordings are sourced from the Core's _default catalogue_, which is stored as physical XLF files (one per translation domain) in `app/Resources/translations/default`. This catalogue is to be filtered by keeping only the translation domains whose name starts with **"Admin"**.

Translations are retrieved from two sources: 

1. The core: XLF translation files for the chosen language, stored in `app/Resources/translations/{locale_code}`, one per domain.
2. The database: if translations were customized previously through this interface.

If the same wording is translated in both core files and database, the database translation is shown.

##### Front office translations

> This translation type has two different subcases:

Reviewer: Are we okay that the front office translations now gathers 1/ the core translations (all frontend wordings), 2/ the theme translations, and 3/ the native modules translations (all front wordings)?

Author: No

Core domains start with "Shop", whereas module domains _end_ with "Shop".

Therefore:

- If the user chooses "Core translations", then it only gets the ones from the core.
- If the user chooses a theme, then it gets all the wordings found in the theme's templates, including any core wording also found there.

This has one caveat: you can't redefine a core translation for a given theme if it's not used in one of the theme's template files.

Example: If a given core wording is displayed in front office, but the wording is physically located in a PHP core file and not within one of the theme's templates, then the wording won't show up for that theme. On the contrary, the wording will show up if it's found in one of the theme's template.

The alternative would be to display all the core wordings, but there's no way to tell if they're actually used in the theme or not.

Reviewer: I quote the summary email sent after the meeting to discuss this feature:

> Autrefois traductions du thème, la seconde entrée s'intitulera désormais traductions du front-office pour regrouper l'ensemble des wordings présents en front, soit les traductions du _core_ (front-end), du _thème_ (ensemble des wordings utilisés dans le template du thème) et des _modules natifs_ (front). [...] À l'export, le comportement reste le même : exporter les traductions du front office reviendra à exporter l'intégralité des catalogues Shop du _core_, Shop des _modules natifs_ et du catalogue des wordings embarqués dans le _thème_.

**i. If the user chooses "Core translations"**

The behavior is exactly like "back office translations", only that translation domains are filtered by domains starting with **"Shop"** instead.

**ii. If the user chooses a theme.**

In this case, wordings are sourced by analyzing all the theme's files. Since this process is slow, the result of this analysis is stored as XLF files in local filesystem cache. If the cache is found, then wordings are read from there. The user can refresh this cache by clicking on the "Find new wordings" button at the top of the page.

Translations are sourced from XLF files within the theme in its `translations/{locale_code}` directory, one per domain. These files may or may not be present.

> User-edited translations are stored in database. To avoid collisions, theme translations must be kept associated to the theme they belong. If a theme redefines a Core translation (ie. same domain, same wording), then the redefined translation should only be used when the FO is using that theme (eg. in case of multishop with different themes for each shop, each theme should use the its own translation even if they both use the same given wording/domain pair).

Author: ```suggestion
User-edited translations are stored in database. To avoid collisions, theme translations must be kept associated to the theme they belong. If a theme redefines a Core translation (ie. same domain, same wording), then the redefined translation should only be used when the FO is using that theme (eg. in case of multishop with different themes for each shop, each theme should use the its own translation even if they both use the same given wording/domain pair). Redefined core translations should only be saved if they're different than the ones defined when not in the theme's context.
```

2. The module: XLF translations files for the chosen language, stored within the module in the `translations/{locale_code}` directory, one per domain.
4. The module's legacy files: PHP translation files for the chosen language, stored within the module in the `translations/` directory, one per language code.
5. The database: if translations were customized previously through this interface.

##### Email translations (subject)

The behavior is exactly like "back office translations", but using only the domain called **"Emails.Subject"**.

##### Email translations (body)

The behavior is exactly like "back office translations", but using only the domain called **"Emails.Body"**.

After saving the translations, the email templates must be regenerated for the translated language.

##### Other translations

The behavior is exactly like "back office translations", but using only the domain called **"messages"**.

## Export translations section

This feature allows the user to export translations in a given language for a given theme. When the user selects the theme/language pair and clicks on "Export", a zip file is generated and downloaded. 

> The generated zip contains a collection of XLF files, one per domain, which is composed of all the front office translations (core + theme). If a core translation is redefined by the theme, then the theme one is used.

Author: What about them?

> The generated zip contains a collection of XLF files, one per domain, which is composed of all the front office translations (core + theme). If a core translation is redefined by the theme, then the theme one is used.

Reviewer: cf. the comment below

> The generated zip contains a collection of XLF files, one per domain, which is composed of all the front office translations (core + theme). If a core translation is redefined by the theme, then the theme one is used.

Author: ```suggestion
The generated zip contains a collection of XLF files, one per domain, which is composed of all the theme's wordings (ie. all wordings found in the theme files), translated to the chosen langue.

Note: Exported translations should be loaded differently to the translation interface. In difference with the translation interface, where non-translated wordings are empty, exported translations should never be empty. If the translation is not defined, use the original wording as translation.
```
